cmd/server/main.go

```diff
@@ -340,6 +342,61 @@ func rateLimitMiddleware(maxReq int, windowLength time.Duration) func(http.Handl
 		}))
 }
 
+func addUserIdentityToContextIfLoggedIn(logger *zap.Logger, db *sqldb.DB) func(http.Handler) http.Handler {
+	fn := func(c context.Context) (context.Context, error) {
+		token, _, err := jwtauth.FromContext(c)
+		if err != nil {
+			return nil, fmt.Errorf("error getting authorization token: %w", err)
+		}
+		if token == nil {
+			return nil, fmt.Errorf("nil authorization token")
+		}
+		emailsClaim, ok := token.PrivateClaims()["emails"]
+		if !ok {
+			return nil, fmt.Errorf("no email claim in token")
+		}
+		emails, ok := emailsClaim.([]interface{})
+		if !ok || len(emails) == 0 {
+			return nil, fmt.Errorf("couldn't find email claim in token: %T", emailsClaim)
+		}
+		// TODO(#18) Handle Multiple Emails in the Token Claims gracefully
+		if len(emails) > 1 {
+			return nil, fmt.Errorf("multiple emails in token: %+v", emails)
+		}
+		email, ok := emails[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("wrong type for email claim: %T", emails[0])
+		}
+		canonical, err := pacta.CanonicalizeEmail(email)
+		if err != nil {
+			return nil, fmt.Errorf("invalid email on token: %q", email)
+		}
+		authnID := token.Subject()
+		if authnID == "" {
+			return nil, fmt.Errorf("couldn't find authn id in jwt")
+		}
+		user, err := db.GetOrCreateUserByAuthn(db.NoTxn(c), pacta.AuthnMechanism_EmailAndPass, authnID, email, canonical)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get user by authn: %w", err)
+		}
+		return jwtauth.WithUserId(c, string(user.ID)), nil
+	}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx, err := fn(r.Context())
```

Comment:
nit: I think this will go away with the above comment, but just in case it doesn't: normally, one only returns a new context from functions that are dedicated to contexts, e.g. `WithX` or `context.WithX` functions. So I'd change it to something like:

```go
fn := func(c context.Context) (pacta.UserID, error) { ... }
[...]
return func(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		uID, err := fn(r.Context())
		ctx := session.WithUserID(ctx, uID)
		next.ServeHTTP(w, r.WithContext(ctx))
	}
}
```

Reply:
This one didn't go away, but I suggested a fix elsewhere

Reply:
Yep - Done.